content/guides/comparators.adoc

```diff
@@ -0,0 +1,635 @@
+        len (min x-len y-len)]
+    (loop [i 0]
+      (if (== i len)
+        ;; If all elements 0..(len-1) are same, shorter vector comes
+        ;; first.
+        (compare x-len y-len)
+        (let [c (cmpf (x i) (y i))]
+          (if (zero? c)
+            (recur (inc i))
+            c))))))
+
+(defn cmp-array-lexi
+  [cmpf x y]
+  (let [x-len (alength x)
+        y-len (alength y)
+        len (min x-len y-len)]
+    (loop [i 0]
+      (if (== i len)
+        ;; If all elements 0..(len-1) are same, shorter array comes
+        ;; first.
+        (compare x-len y-len)
+        (let [c (cmpf (aget x i) (aget y i))]
+          (if (zero? c)
+            (recur (inc i))
+            c))))))
+
+
+(defn cc-cmp
+  [x y]
+  (let [x-cls (comparison-class x)
+        y-cls (comparison-class y)
+        c (compare x-cls y-cls)]
+    (cond (not= c 0) c  ; different classes
+
+          ;; Compare sets to each other as sequences, with elements in
+          ;; sorted order.
+          (= x-cls "clojure.lang.IPersistentSet")
+          (cmp-seq-lexi cc-cmp (sort cc-cmp x) (sort cc-cmp y))
+
+          ;; Compare maps to each other as sequences of [key val]
+          ;; pairs, with pairs in order sorted by key.
+          (= x-cls "clojure.lang.IPersistentMap")
+          (cmp-seq-lexi cc-cmp
+                        (sort-by key cc-cmp (seq x))
+                        (sort-by key cc-cmp (seq y)))
+
+          (= x-cls "java.util.Arrays")
+          (cmp-array-lexi cc-cmp x y)
+
+          ;; Make a special check for two vectors, since cmp-vec-lexi
+          ;; should allocate less memory comparing them than
+          ;; cmp-seq-lexi.  Both here and for comparing sequences, we
+          ;; must use cc-cmp recursively on the elements, because if
+          ;; we used compare we would lose the ability to compare
+          ;; elements with different types.
+          (and (vector? x) (vector? y)) (cmp-vec-lexi cc-cmp x y)
+
+          ;; This will compare any two sequences, if they are not both
+          ;; vectors, e.g. a vector and a list will be compared here.
+          (= x-cls "clojure.lang.Sequential")
+          (cmp-seq-lexi cc-cmp x y)
+
+          :else (compare x y))))
+----
+
+Here is a quick example demonstrating _cc-cmp_'s ability to compare values of different types.
+
+[source,clojure]
+----
+user> (pprint (sort cc-cmp [true false nil Double/MAX_VALUE 10
+                            Integer/MIN_VALUE :a "b" 'c (ref 5)
+                            [5 4 3] '(5 4) (seq [5]) (cons 6 '(1))
+                            #{1 2 3} #{2 1}
+                            {:a 1, :b 2} {:a 1, :b -2}
+                            (object-array [1 2 3 4])]))
+(nil
+ {:a 1, :b -2}
+ {:a 1, :b 2}
+ #{1 2}
+ #{1 2 3}
+ :a
+ #<Ref@1493d9b3: 5>
+ (5)
+ (5 4)
+ [5 4 3]
+ (6 1)
+ c
+ false
+ true
+ -2147483648
+ 10
+ 1.7976931348623157E308
+ "b"
+ [1, 2, 3, 4])
+nil
+----
+
+== TBD
+
+
+TBD: How does Double/NaN compare to other things?
```

Comment:
This is not "TBD", it should sort per http://docs.oracle.com/javase/6/docs/api/java/lang/Double.html#compareTo(java.lang.Double), that is it should sort higher than all other doubles, including +Infinity.